File: js/merge_sort.js
// returns new array, does not mutate original array
function mergeSort(array) {
    if (array.length <= 1) return array;

    let midpoint = Math.floor(array.length / 2);

    let left = mergeSort(array.slice(0, midpoint));
    let right = mergeSort(array.slice(midpoint));

    return merge(left, right);
}

// mutates original array
function mergeSort2(array) {
    if (array.length <= 1) return array;

    let midpoint = Math.floor(array.length / 2);

    let left = mergeSort(array.slice(0, midpoint));
    let right = mergeSort(array.slice(midpoint));

    const temp = merge(left, right);
    for (let i = 0; i < temp.length; i++) {
        array[i] = temp[i];
    }
}

function merge(firstHalf, secondHalf) {
    let sortedArray = [];

    while (firstHalf.length && secondHalf.length) {
        if (firstHalf[0] < secondHalf[0]) {
            sortedArray.push(firstHalf.shift());
        } else {
            sortedArray.push(secondHalf.shift());
        }
    }
    return [...sortedArray, ...firstHalf, ...secondHalf];
}

output.innerHTML += "<h3>testing merge sort in javascript...</h3>";
let unsortedArray = getRandArray(10, 1);
output.innerHTML += `<p>unsorted array -> ${unsortedArray}`;
let sortedArray = mergeSort(unsortedArray);
output.innerHTML += `<p>sorted array -> ${sortedArray}`;

output.innerHTML += "<h3>testing merge sort 2 in javascript...</h3>";
let testArray = getRandArray(10, 1);
output.innerHTML += `<p>unsorted array -> ${testArray}`;
mergeSort2(testArray);
output.innerHTML += `<p>sorted array -> ${testArray}`;

File: python/merge_sort.py
def merge_sort(array):
    if len(array) > 1:
        midpoint = len(array) // 2
        firstHalf = array[:midpoint]
        merge_sort(firstHalf)
        secondHalf = array[midpoint:]
        merge_sort(secondHalf)

        temp = merge(firstHalf, secondHalf)
        for i, val in enumerate(temp):
            array[i] = val
    else:
        return array


def merge(firstHalf, secondHalf):
    mergedList = []
    while len(firstHalf) > 0 and len(secondHalf) > 0:
        if firstHalf[0] < secondHalf[0]:
            mergedList.append(firstHalf[0])
            firstHalf.pop(0)
        else:
            mergedList.append(secondHalf[0])
            secondHalf.pop(0)
    return mergedList + firstHalf + secondHalf

print("testing merge sort in python...")

unsortedArray = [4, 1, 5, 2, 8, 3, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
merge_sort(unsortedArray)
print("sorted array -> " + str(unsortedArray))